processor/src/test/java/org/mapstruct/ap/test/selection/resulttype/ErroneousAmbiguousMostSpecificResultTypeSelectingMapper.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.selection.resulttype;
+
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+/**
+ * @author Filip Hrisafov
+ */
+@Mapper
+public interface ErroneousAmbiguousMostSpecificResultTypeSelectingMapper {
+
+    @Mapping( target = "apple", source = "fruit")
+    AppleFamily map(FruitFamily fruitFamily);
+
+    default GoldenDelicious toGolden(IsFruit fruit) {
+        return fruit != null ? new GoldenDelicious( fruit.getType() ) : null;
+    }
+
+    default Apple toApple1(IsFruit fruit) {
```

Comment:
I saw you used boxed type conversion. Should we have a testcase with the `Number` inheritance somewhere?

Reply:
Did you check `Issue2840Mapper` as part of this PR?